File: educational-core/src/com/jetbrains/edu/learning/checkio/connectors/CheckiOOAuthRestService.java
package com.jetbrains.edu.learning.checkio.connectors;

import com.jetbrains.edu.learning.authUtils.OAuthRestService;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.http.FullHttpRequest;
import io.netty.handler.codec.http.HttpMethod;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.netty.handler.codec.http.QueryStringDecoder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.ide.RestService;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public abstract class CheckiOOAuthRestService extends OAuthRestService {
  private final Pattern myOAuthCodePattern;
  private final CheckiOOAuthConnector myOAuthConnector;

  protected CheckiOOAuthRestService(
    @NotNull String platformName,
    @NotNull String restServicePath,
    @NotNull CheckiOOAuthConnector oauthConnector
  ) {
    super(platformName);
    myOAuthCodePattern = Pattern.compile(restServicePath + "\\?code=(\\w+)");
    myOAuthConnector = oauthConnector;
  }

  @Override
  protected boolean isMethodSupported(@NotNull HttpMethod method) {
    return method == HttpMethod.GET;
  }

  @Override
  protected boolean isHostTrusted(@NotNull FullHttpRequest request) throws InterruptedException, InvocationTargetException {
    final String uri = request.uri();
    final Matcher codeMatcher = myOAuthCodePattern.matcher(uri);
    if (request.method() == HttpMethod.GET && codeMatcher.matches()) {
      return true;
    }
    return super.isHostTrusted(request);
  }

  @Nullable
  @Override
  public String execute(
    @NotNull QueryStringDecoder decoder,
    @NotNull FullHttpRequest request,
    @NotNull ChannelHandlerContext context
  ) throws IOException {
    final String uri = decoder.uri();

    if (myOAuthCodePattern.matcher(uri).matches()) {
      final String code = getStringParameter("code", decoder);
      assert code != null; // cannot be null because of pattern

      LOG.info(myPlatformName + ": OAuth code is handled");
      final String errorMessage = myOAuthConnector.codeHandler(code);

      if (errorMessage == null) {
        return sendOkResponse(request, context);
      }
      else {
        return sendErrorResponse(request, context, errorMessage);
      }
    }

    RestService.sendStatus(HttpResponseStatus.BAD_REQUEST, false, context.channel());
    return "Unknown command: " + uri;
  }
}

File: educational-core/src/com/jetbrains/edu/learning/checkio/connectors/CheckiOOAuthRestService.kt
package com.jetbrains.edu.learning.checkio.connectors

import com.intellij.util.io.origin
import com.jetbrains.edu.learning.authUtils.OAuthRestService
import com.jetbrains.edu.learning.checkio.utils.CheckiONames
import com.jetbrains.edu.learning.checkio.utils.CheckiONames.CHECKIO_URL
import com.jetbrains.edu.learning.courseFormat.EduFormatNames.CHECKIO
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.http.*
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.util.regex.Pattern

abstract class CheckiOOAuthRestService(platformName: String, private val oAuthConnector: CheckiOOAuthConnector) :
  OAuthRestService(platformName) {
  private val oAuthCodePattern: Pattern = oAuthConnector.getOAuthPattern()

  override fun getServiceName(): String = oAuthConnector.serviceName

  @Throws(InterruptedException::class, InvocationTargetException::class)
  override fun isHostTrusted(request: FullHttpRequest, urlDecoder: QueryStringDecoder): Boolean {
    return if (request.method() === HttpMethod.GET
               // If isOriginAllowed is `false` check if it is a valid oAuth request with empty origin
               && (isOriginAllowed(request) === OriginCheckResult.ALLOW || oAuthConnector.isValidOAuthRequest(request, urlDecoder))) {
      true
    }
    else super.isHostTrusted(request, urlDecoder)
  }

  @Throws(IOException::class)
  override fun execute(
    urlDecoder: QueryStringDecoder,
    request: FullHttpRequest,
    context: ChannelHandlerContext
  ): String? {
    val uri = urlDecoder.uri()
    LOG.info("Request: $uri")
    if (oAuthCodePattern.matcher(uri).matches()) {
      // cannot be null because of pattern
      val code = getStringParameter(CODE_ARGUMENT, urlDecoder)!!
      LOG.info("$platformName: OAuth code is handled")
      val success = oAuthConnector.login(code)
      return if (success) {
        sendOkResponse(request, context)
      }
      else {
        val errorMessage = "Failed to login to $CHECKIO"
        sendErrorResponse(request, context, errorMessage)
      }
    }
    sendStatus(HttpResponseStatus.BAD_REQUEST, false, context.channel())
    return "Unknown command: $uri"
  }

  override fun isOriginAllowed(request: HttpRequest): OriginCheckResult {
    val originAllowed = super.isOriginAllowed(request)
    if (originAllowed == OriginCheckResult.FORBID) {
      val origin = request.origin ?: return OriginCheckResult.FORBID
      return if (origin == CHECKIO_URL) OriginCheckResult.ALLOW
      else OriginCheckResult.ASK_CONFIRMATION
    }
    return originAllowed
  }
}
